File: src/Services/Notification/Notification.Domain/Entities/Notification.cs
using BuildingBlocks.Domain.Exceptions;
using Notification.Domain.Enums;

namespace Notification.Domain.Entities;

public class Notification : AggregateRoot
{
    public string UserId { get; private set; } = string.Empty;
    public string Username { get; private set; } = string.Empty;
    public NotificationType Type { get; private set; }
    public string Title { get; private set; } = string.Empty;
    public string Message { get; private set; } = string.Empty;

    public string? HtmlContent { get; private set; }
    public string Data { get; private set; } = string.Empty;
    public NotificationStatus Status { get; private set; } = NotificationStatus.Unread;
    public ChannelType Channels { get; private set; } = ChannelType.InApp;
    public DateTimeOffset? ReadAt { get; private set; }
    public Guid? AuctionId { get; private set; }
    public Guid? BidId { get; private set; }
    public Guid? OrderId { get; private set; }
    public string? ReferenceId { get; private set; }

    public static Notification Create(
        string userId,
        string username,
        NotificationType type,
        string title,
        string message,
        ChannelType channels = ChannelType.InApp,
        string? data = null,
        Guid? auctionId = null,
        Guid? bidId = null,
        Guid? orderId = null,
        string? referenceId = null,
        string? htmlContent = null)
    {
        return new Notification
        {
            Id = Guid.NewGuid(),
            UserId = userId,
            Username = username ?? string.Empty,
            Type = type,
            Title = title,
            Message = message,
            Channels = channels,
            Data = data ?? string.Empty,
            HtmlContent = htmlContent,
            Status = NotificationStatus.Pending,
            AuctionId = auctionId,
            BidId = bidId,
            OrderId = orderId,
            ReferenceId = referenceId,
            CreatedAt = DateTimeOffset.UtcNow
        };
    }

    public void MarkAsSent()
    {
        if (Status == NotificationStatus.Pending)
            Status = NotificationStatus.Unread;
    }

    public void MarkAsRead()
    {
        if (Status == NotificationStatus.Dismissed)
            throw new InvalidEntityStateException(nameof(Notification), nameof(NotificationStatus.Dismissed), "Cannot mark dismissed notification as read");

        Status = NotificationStatus.Read;
        ReadAt = DateTimeOffset.UtcNow;
    }

    public void MarkAsUnread()
    {
        if (Status == NotificationStatus.Dismissed)
            throw new InvalidEntityStateException(nameof(Notification), nameof(NotificationStatus.Dismissed), "Cannot mark dismissed notification as unread");

        Status = NotificationStatus.Unread;
        ReadAt = null;
    }

    public void Dismiss()
    {
        Status = NotificationStatus.Dismissed;
    }

    public void Archive()
    {
        if (Status == NotificationStatus.Pending)
            throw new InvalidEntityStateException(nameof(Notification), nameof(NotificationStatus.Pending), "Cannot archive pending notification");

        Status = NotificationStatus.Archived;
    }

    public bool IsRead => Status == NotificationStatus.Read;
    public bool IsDismissed => Status == NotificationStatus.Dismissed;
    public bool IsArchived => Status == NotificationStatus.Archived;
    public bool IsPending => Status == NotificationStatus.Pending;
}

File: Web/types/notification.ts
export interface Notification {
  id: string;
  userId: string;
  username?: string;
  type: NotificationType;
  title: string;
  message: string;
  htmlContent?: string;
  data: string;
  status: NotificationStatus;
  channels?: string;
  readAt?: string;
  auctionId?: string;
  bidId?: string;
  orderId?: string;
  createdAt: string;
}

export enum NotificationType {
  AuctionCreated = "AuctionCreated",
  AuctionUpdated = "AuctionUpdated",
  AuctionDeleted = "AuctionDeleted",
  AuctionFinished = "AuctionFinished",
  AuctionStarted = "AuctionStarted",
  AuctionEnding = "AuctionEnding",
  AuctionEndingSoon = "AuctionEndingSoon",
  AuctionLost = "AuctionLost",
  BidPlaced = "BidPlaced",
  BidAccepted = "BidAccepted",
  BidRejected = "BidRejected",
  Outbid = "Outbid",
  OutBid = "OutBid",
  AuctionWon = "AuctionWon",
  BuyNowExecuted = "BuyNowExecuted",
  OrderCreated = "OrderCreated",
  OrderShipped = "OrderShipped",
  OrderDelivered = "OrderDelivered",
  PaymentReceived = "PaymentReceived",
  PaymentCompleted = "PaymentCompleted",
  ReviewReceived = "ReviewReceived",
  WelcomeEmail = "WelcomeEmail",
  PasswordReset = "PasswordReset",
  AccountVerification = "AccountVerification",
  System = "System"
}

export enum NotificationStatus {
  Pending = "Pending",
  Sent = "Sent",
  Delivered = "Delivered",
  Failed = "Failed",
  Unread = "Unread",
  Read = "Read",
  Dismissed = "Dismissed"
}

export enum ChannelType {
  None = 0,
  InApp = 1,
  Email = 2,
  Sms = 4,
  Push = 8
}

export interface NotificationSummary {
  unreadCount: number;
  totalCount: number;
  recentNotifications: Notification[];
}

export interface PagedNotifications {
  items: Notification[];
  totalCount: number;
  pageNumber: number;
  pageSize: number;
  totalPages: number;
}

export interface BroadcastNotificationDto {
  type: NotificationType;
  title: string;
  message: string;
  targetRole?: string;
}

export interface NotificationStats {
  totalNotifications: number;
  unreadNotifications: number;
  todayCount: number;
  byType: Record<string, number>;
}
